libsolidity/analysis/OverrideChecker.cpp

```diff
@@ -897,10 +897,11 @@ OverrideChecker::OverrideProxyBySignatureMultiSet const& OverrideChecker::inheri
 				if (var->isPublic())
 					functionsInBase.emplace(OverrideProxy{var});
 
-			for (OverrideProxy const& func: inheritedFunctions(*base))
-				functionsInBase.insert(func);
-
 			result += functionsInBase;
+
+			for (OverrideProxy const& func: inheritedFunctions(*base))
+				if (!functionsInBase.count(func))
+					result.insert(func);
```

Comment:
> The insidious part of the change is that ``functionsInBase`` is a set and ``result`` is a multiset...
>
> So before the change, the inner loop over ``inheritedFunctions`` potentially ranged over multiple copies of functions, but only one would end up in ``functionsInBase``. 

> in ``Impl`` when looking at functions inherited from ``Ext``, it'd only consider one base function ``foo`` - and it happened to be the one in the interface. And since we don't require ``override`` for interface functions it didn't complain.
>
> With the change I propagate *all* function with the signature from ``Ext`` to ``Impl``. Which means I need to override both in ``Impl`` - which is a bit weird, but I think that's how our spec works right now.

Thanks for this explanation. So I actually interpreted this wrong the first time I looked at it. Now it makes sense. I did a deeper dive through the code now and I think I finally understand well what is happening and why.

I think that this lets through only a single function on purpose. If you inherit from `A`, `B` and `C`, each of them must have already settled on a single definition for `foo()`. So the code above should never have to insert more than one function into `result`.

The bug was introduced in #11628, specifically by https://github.com/ethereum/solidity/commit/f7916f2940d3266a28c001a08a474f0c6eb3df01#diff-980e91f229ba8ff7a7ed8705977e094851571ba47f544a50a9722371e8086dc4L521-R522. It wrongly assumes that it always gets that single, reconciled `foo()` from a parent contract that you would end up calling if you referred to `foo()` inside that parent contract.

So to fix this cleanly, `inheritedFunctions()` has to satisfy this assumption. Basically, it needs to perform the same lookup logic you get when you call `foo()`. It should use `CallableDeclaration::resolveVirtual()`. The question is - can we do it here? That function might depend on something that's not yet available at this analysis stage.

If we can't do that, a viable workaround would be just to return a non-interface function if one is available.

Reply:
As for testing this, I think that we just need to cover variants of this "bottleneck" in the inheritance hierarchy:
```
A   B
 \ /
  X
  |
  Y
```
the problem should really exist only if `X` does not have its own version of the function and there are multiple versions in its ancestors.

Reply:
The bug was already in the initial implementation of overrides and is older than https://github.com/ethereum/solidity/pull/11628 - https://github.com/ethereum/solidity/pull/11628 only made it more visible, since you end up *not* having to override stuff due to this - but in situations like https://github.com/ethereum/solidity/pull/12616#discussion_r799823373, if ``X`` does not implement ``f``, but ``A``, ``B`` and ``C`` do, we've always only considered ``A.f`` for the override logic, which has always been wrong.

Reply:
```
abstract contract I { function f() external virtual; }
abstract contract A is I {
}
abstract contract B is I {
        function f() public pure virtual override {}
}
contract X is A, B {
}

contract Y is X {
        function f() public pure override(B) {}
}
```
This for example fails on 0.8.11 - it wants you to use ``override(I)`` instead of ``B``. Which is wrong :-). And once you extend it to more than one base this error becomes visible... that can only happen, if you have a common base in which the function is unimplemented and then multiple paths without mention of the function and one path that defines it... that's the situation in which we ended up *not* requiring an ``override`` in ``X``... unfortunately, we did not clearly define what should happen in ``Y`` in this situation - respectively, we implemented it incorrectly...

Reply:
This is the relevant test case in zeppelin, let me have another look at that:
https://app.circleci.com/pipelines/github/ethereum/solidity/21931/workflows/31931f3e-fc90-4ebe-8670-f57eeb69d6a7/jobs/961852

Reply:
Reducing the case of ``state()`` in there to only the relevant parts, then it's this situation:
```
interface IG {
	function f() external pure virtual;
}

abstract contract G is IG {
	function f() public pure virtual override {}
}
abstract contract GS is IG {
}

abstract contract GTL is IG {
	function f() public pure virtual override {}
}

abstract contract GVC is G {
}

abstract contract GCB is G {
	
}

contract MD is GCB, GS, GTL, GVC {
	function f() public pure override (IG, G, GTL) {}
}
```

We *do* currently require ``IG`` in the override list. The same happens if you change ``IG`` to an abstract contract.
The situation is a bit different than I thought. I think the third override requirement is due to ``GVC`` and ``GCB`` both bringing in an *implemented* version of ``f``... however I'm not quite sure why it is ``IG`` of all things the compiler asks to additionally override here...

Reply:
Ok, maybe zeppelin breaking was actually only a bug in my preliminary fix in the first commit, not an issue with implementing https://github.com/ethereum/solidity/pull/12616#discussion_r797985635...

Reply:
> If we can't do that, a viable workaround would be just to return a non-interface function if one is available.

That's in any case what the first commit in this PR does, isn't it? And that commit breaks zeppelin... I'm still trying to wrap my head around this entirely, though :-).